### PR TITLE
Add better warning messages librz/bin/format/elf/*

### PR DESCRIPTION
**Pull request by ghost**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

* ~~I need help with one failing test (db/archos/linux-x64/dbg_oo ood check for ptrace errors)~~
* ~~I used RZ_LOG_WARN everywhere, not sure if it is a good idea.~~
* ~~There is 2 types of warning: memory and logic. Is there a specific function to report out of memory error (malloc, sdb, rz_str, ht_uu, rz_vector, etc).~~
* Warnings message should be review (grammar and logic)

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by wargio** [comment]
RZ_LOG_WARN is ok

File: librz/bin/format/elf/elf_dynamic.c [hunk 1]
```diff
@@ -16,80 +16,105 @@ static bool get_dynamic_entry(ELFOBJ *bin, ut64 offset, Elf_(Dyn) * entry) {
...
+		RZ_LOG_WARN("Dynamic section allocation failed.\n");
+		Elf_(rz_bin_elf_dt_dynamic_free)(result);
 		return NULL;
 	}
 
 	result->dt_needed = rz_vector_new(sizeof(Elf_(Word)), NULL, NULL);
 	if (!result->dt_needed) {
-		ht_uu_free(result->info);
-		free(result);
+		RZ_LOG_WARN("Dynamic section allocation failed.\n");
```
Comment (wargio) [inline librz/bin/format/elf/elf_dynamic.c]: here use RZ_LOG_ERROR. if it is a fail, then is ERROR, like for example when you fail to alloc.
When is a problem, like you expect to have 255 max values but you have 1000, then is a warning.

**Review by wargio — CHANGES_REQUESTED** [review]

File: librz/bin/format/elf/elf_dynamic.c (hunk 1 above)
Comment (wargio) [reply librz/bin/format/elf/elf_dynamic.c]: same for all failed allocs in this code.

Comment (ghost) [reply librz/bin/format/elf/elf_dynamic.c]: Ty

File: librz/bin/format/elf/elf_sections.c [hunk 2]
```diff
@@ -76,12 +76,12 @@ static bool create_section_from_phdr(ELFOBJ *bin, RzVector *result, const char *
 	section.size = sz;
 	section.name = strdup(name);
 	if (!section.name) {
-		RZ_LOG_WARN("Failed to allocate memory for the section name.")
+		RZ_LOG_WARN("Section name allocation railed.\n")
```
Comment (thestr4ng3r) [inline librz/bin/format/elf/elf_sections.c]:
```suggestion
		RZ_LOG_WARN("Section name allocation failed.\n")
```

File: librz/bin/format/elf/elf_symbols.c [hunk 3]
```diff
@@ -512,7 +515,19 @@ Elf_(Word) Elf_(rz_bin_elf_get_number_of_dynamic_symbols)(RZ_NONNULL ELFOBJ *bin
 		return result;
 	}
 
-	return get_number_of_symbols_from_heuristic(bin);
+	result = get_number_of_symbols_from_section(bin);
+	if (result) {
+		RZ_LOG_WARN("The number of dynamic symbols is deduced from section information.\n");
```
Comment (thestr4ng3r) [inline librz/bin/format/elf/elf_symbols.c]: This by itself doesn't say much when appearing to the user. Before the `result = get_number_of_symbols_from_section(bin);` line I would do something like
```c
RZ_LOG_WARN("Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...\n");
```

And then in the end if all heuristics have failed:
```c
RZ_LOG_WARN("Failed to determine the number of dynamic symbols from heuristics.\n");
```
(These log strings are just examples, you will probably know better how to formulate them accurately here)

**Review by thestr4ng3r — COMMENTED** [review]
> There is 2 types of warning: memory and logic. Is there a specific function to report out of memory error (malloc, sdb, rz_str, ht_uu, rz_vector, etc).

Almost all memory errors should simply be propagated outwards by returning `NULL` or whatever is appropriate in the specific case, without printing any warning, which is pretty much the standard way to do this in C. In particular, mallocs that allocate a static amount of memory like `rz_list_new()` shouldn't have any warnings at all.

A different case is when you allocate memory of a size given from the outside though, for example allocating an entire block of the a segment size that comes from the binary. Then it can make sense to print a warning to indicate there is something wrong with the binary or you just don't have enough ram to deal with that binary.

> Warnings message should be review (grammar and logic)

General comment for all `read failed` and `parsing failed`: Instead of e.g. `Note prstatus parsing failed.` I would write `Failed to parse NT_PRSTATUS note.`. This makes it a bit more comprehensible imo.

File: librz/bin/format/elf/elf_symbols.c (hunk 3 above)
Comment (wargio) [reply librz/bin/format/elf/elf_symbols.c]: i agree on this one, but use `RZ_LOG_ERROR` for `RZ_LOG_WARN("Failed to determine the number of dynamic symbols from heuristics.\n");`.

**Comment by ghost** [comment]
> > There is 2 types of warning: memory and logic. Is there a specific function to report out of memory error (malloc, sdb, rz_str, ht_uu, rz_vector, etc).
> 
> Almost all memory errors should simply be propagated outwards by returning `NULL` or whatever is appropriate in the specific case, without printing any warning, which is pretty much the standard way to do this in C. In particular, mallocs that allocate a static amount of memory like `rz_list_new()` shouldn't have any warnings at all.
> 
> A different case is when you allocate memory of a size given from the outside though, for example allocating an entire block of the a segment size that comes from the binary. Then it can make sense to print a warning to indicate there is something wrong with the binary or you just don't have enough ram to deal with that binary.
> 
> > Warnings message should be review (grammar and logic)
> 
> General comment for all `read failed` and `parsing failed`: Instead of e.g. `Note prstatus parsing failed.` I would write `Failed to parse NT_PRSTATUS note.`. This makes it a bit more comprehensible imo.

How far the propagation should go?

**Comment by ghost** [comment]
~~I need help with db/archos/linux-x64/dbg_oo, I can't figure out how to fix the test because of `REGEXP_FILTER_ERR=([a-zA-Z-]+\s+)`.~~

File: librz/bin/format/elf/elf_symbols.c (hunk 3 above)
Comment (ghost) [reply librz/bin/format/elf/elf_symbols.c]: Okay ty

**Comment by XVilka** [comment]
```
[XX] db/formats/elf/main ELF: endbr-main-mov
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?v main' bins/elf/endbr-main
-- stderr
--- expected
+++ actual
@@ -1,3 +1,0 @@
-WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
-WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
-WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
```

**Comment by ghost** [comment]
@XVilka It is really weird.

* MacOs doesn't failed
* Freebsd doesn't failed
* Openbsd doesn't failed
* gcc-asan doesn't failed on a related test (some PE test failed)
* gcc and clang failed because no WARNING was printed.

In theory there shouldn't be any side effect because gcc-asan didn't failed. I am confused. I don't know where is the problem. I couldn't reproduce the failed test on my computer with and without asan.

**Review by XVilka — APPROVED** [review]
ASAN and Debian Wheezy failures are unrelated

**Review by wargio — APPROVED** [review]

**Pull request merged**